util/auto_tune_compressor.cc

```diff
@@ -144,138 +152,222 @@ CostAwareCompressor::CostAwareCompressor(const CompressionOptions& opts)
+                                  compressed_output, out_compression_type,
+                                  local_wa);
+  }
 }
 
-Compressor::ManagedWorkingArea CostAwareCompressor::ObtainWorkingArea() {
-  auto wrap_wa = allcompressors_.back().back()->ObtainWorkingArea();
+Compressor::ManagedWorkingArea AutoTuneCompressor::ObtainWorkingArea() {
+  auto wrap_wa = compressors_.back()->ObtainWorkingArea();
   auto wa = new CostAwareWorkingArea(std::move(wrap_wa));
   // Create cost predictors for each compression type and level
-  wa->cost_predictors_.reserve(allcompressors_.size());
-  for (size_t i = 0; i < allcompressors_.size(); i++) {
-    CompressionType type = static_cast<CompressionType>(i + 1);
-    if (allcompressors_[type - 1].size() == 0) {
-      wa->cost_predictors_.emplace_back();
-      continue;
-    } else {
-      std::vector<IOCPUCostPredictor*> predictors_diff_levels;
-      predictors_diff_levels.reserve(kCompressionLevels[type - 1].size());
-      for (size_t j = 0; j < kCompressionLevels[type - 1].size(); j++) {
-        predictors_diff_levels.emplace_back(new IOCPUCostPredictor(10));
-      }
-      wa->cost_predictors_.emplace_back(std::move(predictors_diff_levels));
+  wa->cost_predictors_.reserve(compressors_.size());
+  for (size_t i = 0; i < kCompressionLevels.size(); i++) {
+    for (size_t j = 0; j < kCompressionLevels[i].size(); j++) {
+      wa->cost_predictors_.emplace_back(new IOCPUCostPredictor(kWindow));
     }
   }
   return ManagedWorkingArea(wa, this);
 }
-void CostAwareCompressor::ReleaseWorkingArea(WorkingArea* wa) {
+void AutoTuneCompressor::MeasureUtilization() { usage_tracker_.Record(); }
+void AutoTuneCompressor::ReleaseWorkingArea(WorkingArea* wa) {
   // remove all created cost predictors
-  for (auto& prdictors_diff_levels :
+  for (auto& predictor :
        static_cast<CostAwareWorkingArea*>(wa)->cost_predictors_) {
-    for (auto& predictor : prdictors_diff_levels) {
-      delete predictor;
-    }
+    delete predictor;
   }
   delete static_cast<CostAwareWorkingArea*>(wa);
 }
+size_t AutoTuneCompressor::SelectCompressionBasedOnIOGoalCPUBudget(
+    CostAwareWorkingArea* wa) {
+  // If no budgets are available, use default choice
+  if (!cpu_budget_ || !io_goal_) {
+    size_t default_choice = 0;
+    return default_choice;
+  }
+  if ((block_count_.fetch_add(1, std::memory_order_relaxed)) %
+          kDecideEveryNBlocks !=
+      0) {
+    return cur_compressor_idx_;
+  }
+  MeasureUtilization();
+  auto cpu_util = usage_tracker_.GetCpuUtilization();
+  auto io_util = usage_tracker_.GetIoUtilization();
+  TEST_SYNC_POINT_CALLBACK("AutoTuneCompressorWrapper::SetCPUUsage", &cpu_util);
+  TEST_SYNC_POINT_CALLBACK("AutoTuneCompressorWrapper::SetIOUsage", &io_util);
+  // Get available budgets
+  auto cpu_goal = cpu_budget_->GetMaxRate();
+  auto io_goal = io_goal_->GetMaxRate();
+  auto cpu_mingoal = cpu_budget_->GetMinRate();
+  auto io_mingoal = io_goal_->GetMinRate();
```

Comment:
Please also update the variable names to use words like upper_bound, lower_bound consistently